src/java/org/apache/cassandra/service/reads/ReadCallback.java

```diff
@@ -56,29 +59,104 @@
 public class ReadCallback<E extends Endpoints<E>, P extends ReplicaPlan.ForRead<E>> implements RequestCallback<ReadResponse>
 {
     protected static final Logger logger = LoggerFactory.getLogger( ReadCallback.class );
-    private class WarningCounter
+
+    private class WarnAbortCounter
     {
-        // the highest number of tombstones reported by a node's warning
-        final AtomicInteger tombstoneWarnings = new AtomicInteger();
-        final AtomicInteger maxTombstoneWarningCount = new AtomicInteger();
-        // the highest number of tombstones reported by a node's rejection. This should be the same as
-        // our configured limit, but including to aid in diagnosing misconfigurations
-        final AtomicInteger tombstoneAborts = new AtomicInteger();
-        final AtomicInteger maxTombstoneAbortsCount = new AtomicInteger();
-
-        // TODO: take message as arg and return boolean for 'had warning' etc
-        void addTombstoneWarning(InetAddressAndPort from, int tombstones)
+        final AtomicInteger warnings = new AtomicInteger();
+        // the highest number reported by a node's warning
+        final AtomicLong maxWarningCount = new AtomicLong();
+
+        final AtomicInteger aborts = new AtomicInteger();
+        // the highest number reported by a node's rejection.
+        final AtomicLong maxAbortsCount = new AtomicLong();
+
+        void addWarning(InetAddressAndPort from, long value)
         {
             if (!waitingFor(from)) return;
```

Comment:
I think we may also want to change this to use `assertWaitingFor()` rather than just silently doing nothing if we get a response from an unexpected replica. (I missed this in CASSANDRA-16807.)

EDIT: We actually may not need this at all, given `onResponse()` already calls `assertWaitingFor()`.

Reply:
That would also allow us to make `WarnAbortCounter` a static class.